src/lib/redux/customerData.ts

```diff
@@ -21,9 +36,76 @@ export const setCurrentCustomerIdInRedux = (customerId: string): void => {
-    store.getState, 
+    selectCurrentCustomerId,
+    store.getState,
     (currentCustomerId, state) => selectCustomerById(state, currentCustomerId));
+
+export const selectMeterReadingsByCustomerId = createSelector(
+    selectAllMeterReadingsArray,
+    getCustomerId,
+    (allMeterReadings, customerId) => allMeterReadings?.filter(meterReading => meterReading.customerId === customerId)
+)
+
+export const selectPaymentsByCustomerId = createSelector(
+    selectAllPaymentsArray,
+    getCustomerId,
+    (allPayments, customerId) => allPayments?.filter(payment => payment.billedToId === customerId)
+)
+
+// TODO: we might have to change this once we factor in grace periods
+export const selectMeterReadingsInCurrentPeriod = createSelector(
+    selectAllMeterReadingsArray,
+    (allMeterReadings) => allMeterReadings.filter(meterReading => !isBeforeCurrentPeriod(meterReading.date))
+)
+
+// TODO: we might have to change this once we factor in grace periods
+export const selectPaymentsInCurrentPeriod = createSelector(
+    selectAllPaymentsArray,
+    (payments) => payments.filter(payment => !isBeforeCurrentPeriod(payment.date))
+)
+
+export const selectCustomersToMeter = createSelector(
+    selectMeterReadingsInCurrentPeriod,
+    selectAllCustomersArray,
+    (allMeterReadingsInCurrentPeriod, allCustomers) => {
+        const meteredCustomerIds = new Set(allMeterReadingsInCurrentPeriod.map(meterReading => meterReading.customerId));
+        return allCustomers.filter(customer => !meteredCustomerIds.has(customer.id))
+    }
+)
+
+// For each customer who has been metered this period, 
+// we calculate the amount
+// the amount they need to pay
+export const selectCustomersToCollect = createSelector(
+    selectMeterReadingsInCurrentPeriod,
+    selectPaymentsInCurrentPeriod,
+    selectCustomersToMeter,
+    selectAllCustomersArray,
+    (periodMeterReadings, periodPayments, unmeteredCustomers, allCustomers) => {
+        const unmeteredCustomerIds = new Set(unmeteredCustomers.map((customer: CustomerRecord) => customer.id));
+        const customersToCollect: CustomerRecord[] = []
+
+        allCustomers
+            .filter((customer: CustomerRecord) => !unmeteredCustomerIds.has(customer.id))
+            .forEach((customer: CustomerRecord) => {
+                const amountOwedThisPeriod = periodMeterReadings
+                    .filter(meterReading => meterReading.customerId === customer.id)
+                    .reduce((totalAmountOwed, meterReading: MeterReadingRecord) => totalAmountOwed + meterReading.amountBilled, 0);
+
+                const amountPaidThisPeriod = periodPayments
+                    .filter(payments => payments.billedToId === customer.id)
+                    .reduce((totalAmountPaid, payment: PaymentRecord) => totalAmountPaid + payment.amount, 0);
+
+                if (amountOwedThisPeriod <= amountPaidThisPeriod) {
+                    customersToCollect.push(customer)
+                }
```

Comment:
Since this is such a fat selector, I'm wondering if you could extract some parts of this out to helper functions, like maybe what's happening in this forEach? Although that may end up negating the performance benefits of selectors.

Reply:
Good idea, I was afraid that this function might be hard to follow. Created selectors to calculate `amountOwedThisPeriod` and `amountPaidThisPeriod`.